src/v/cloud_topics/reconciler/reconciler.cc

```diff
@@ -102,6 +106,64 @@ void reconciler::detach(const model::ntp& ntp) {
     }
 }
 
+uint64_t reconciler::compute_max_offset_lag() const {
+    uint64_t max_lag = 0;
+    for (const auto& [_, src] : _sources) {
+        auto lso = src->last_stable_offset();
+        if (lso == kafka::offset{}) {
+            // LSO not available right now.
+            continue;
+        }
+        auto lro = std::max(src->last_reconciled_offset(), kafka::offset{0});
+
+        if (lso >= lro) {
+            // LSO < LRO is impossible. Not metrics's job to enforce it.
+            auto lag = static_cast<uint64_t>((lso - lro)());
+            max_lag = std::max(max_lag, lag);
+        }
+
+        if (ss::need_preempt()) {
```

Comment:
i think we should drop the metrics that use `need_preempt` and rethink how we can collect this information. it's often (not always) the case that an equivalently good metric can be either continuously computed inline with normal operation and metric collect samples this value, or if that is hard, then a background fiber can compute the value periodically and the metric collect can report a slightly out-of-date value.

we don't need to figure out that alternative in this pr, but i do think we should remove these loops from metric collection hot path.

as far as the metric itself goes, would we want to collect this lag value at the CTP level? For example, a CTP knows its own last reconciled offset and it's own max offset, so would that be enough?